### PR TITLE
feat/seg branch: segmentation layer implementation and correction on typos

### PR DESCRIPTION
**Pull request by Jeongwook-Lee**

### CODE REVIEW

File: dataset/dataset.py [hunk 1]
```diff
@@ -188,13 +186,18 @@ def __getitem__(self, indx):
...
-        # Convert int64 to uint8
+        
+        x_mask_ref = _binarize(x_mask_ref)
+        x_mask_test = _binarize(x_mask_test)
+
+        # Convenrt datatypes
+        x_ref = x_ref.astype(np.float32)
+        x_test = x_test.astype(np.float32)
+        x_mask_ref = x_mask_ref.astype(np.uint8)
+        x_mask_test = x_mask_test.astype(np.uint8)
```
Comment (KevinTheRainmaker) [inline dataset/dataset.py:197]: x_mask_ref와 x_mask_test에 대해서 _binarize와 astype을 해주는 이유가 무엇인가요? 하더라고 line 188 이전에 해야하는 것 아닌가해서 여쭤봅니다

File: dataset/dataset.py [hunk 2]
```diff
@@ -224,20 +227,17 @@ def __getitem__(self, indx):
...
-
+        '''
         # Data augmentation in case of training:
         if self._mode == "train":
-            x_ref, x_test, x_mask = self._augment(
-                x_ref, x_test, x_mask)
-
-        # Change datatype from bool to float
-        x_ref, x_test, x_mask = x_ref.astype(np.float), x_test.astype(
-            np.float), x_mask.astype(np.float)
```
Comment (KevinTheRainmaker) [inline dataset/dataset.py:235]: _augment의 출력이 float인가요? 이 부분 제거하기 전에 확인하셨는지 궁금합니다.

File: models/layers.py [hunk 3]
```diff
@@ -66,10 +71,10 @@ def __init__(
         widths: List[int],
         decoder_channels: int,
         scale_factors: List[int],
-        num_classes: int = 2,
+        num_classes: int = 1,
```
Comment (KevinTheRainmaker) [inline models/layers.py:74]: binary prediction이면 class 2개 아닌가요..?

File: training.py [hunk 4]
```diff
@@ -107,37 +109,49 @@ def train(
...
-        for (reference, testimg), mask in tqdm(dataset_train):
+        weight = [0.5, 0.5]
+        for (reference, testimg), (mask_ref, mask_test, mask) in tqdm(dataset_train):
             # Reset the gradients:
             optimizer.zero_grad()
 
             # Loss gradient descend step:
-            it_loss = evaluate(reference, testimg, mask)
+            it_change_loss, it_segmentation_loss = evaluate(reference, testimg, mask_ref, mask_test, mask)
+            weight[0] = it_segmentation_loss/(it_change_loss + it_segmentation_loss)
```
Comment (KevinTheRainmaker) [inline training.py:152]: 특이한 loss 식이네요. 레퍼런스가 있나요?

File: training.py [hunk 5]
```diff
@@ -182,9 +196,13 @@ def validation_phase(epc):
         epoch_loss_eval = 0.0
         tool4metric.clear()
         with torch.no_grad():
-            for (reference, testimg), mask in dataset_val:
-                epoch_loss_eval += evaluate(reference,
-                                            testimg, mask).to("cpu").numpy()
+            weight = [0.5, 0.5]
+            for (reference, testimg), (mask_ref, mask_test, mask) in dataset_val:
+                it_change_loss, it_segmentation_loss = evaluate(reference, testimg, mask_ref, mask_test, mask)
+                weight[0] = it_segmentation_loss/(it_change_loss + it_segmentation_loss)
```
Comment (KevinTheRainmaker) [inline training.py:202]: 상동

File: dataset/dataset.py (hunk 1 above)
Comment (Jeongwook-Lee) [reply dataset/dataset.py:197]: mask 값들을 0,1로 binarize하기 위해서 해주었습니다. 추가적으로 astype은 용진님께서 mask 타입을 uint8로 하는 것이 좋다고 하셔서 적용했습니다. 결과는 동일할 것 같긴 합니다만 line 188이전에 하는 것이 맞는 것 같습니다.

File: models/layers.py (hunk 3 above)
Comment (Jeongwook-Lee) [reply models/layers.py:74]: 저도 그렇게 생각하긴 하는데 최종 segmentation mask output을 one channel로 출력해야되지 않나요..??

File: training.py (hunk 4 above)
Comment (Jeongwook-Lee) [reply training.py:152]: segmentation loss 변화에 따라 정확이 어떤 값으로 파라미터 값을 맞춰야하는지 잘 모르겠어서 glunet 논문에서 adaptive  loss에서 두 loss 를 동일하게 값을 맞춰주길래 이 부분은 여기서 가져왔습니다. 사실 cd 논문은 아니여서 이부분을 따라해도 될지는 모르겠습니다. (https://arxiv.org/pdf/2104.03308.pdf  -> 5page)

File: training.py (hunk 5 above)
Comment (Jeongwook-Lee) [reply training.py:202]: 상동

File: dataset/dataset.py (hunk 2 above)
Comment (Jeongwook-Lee) [reply dataset/dataset.py:235]: segmenration gt mask도 augmentation 적용하는 것으로 233, 234줄에 수정해두었습니다.

File: models/layers.py (hunk 3 above)
Comment (Jeongwook-Lee) [reply models/layers.py:74]: cd 부분에서도 binary prediction이지만 최종 one channel로 한 것 같아서 동일하게 했습니다.

File: dataset/dataset.py (hunk 2 above)
Comment (KevinTheRainmaker) [reply dataset/dataset.py:235]: 음 아뇨 그 부분은 확인했습니다만, line 233에서 _augment로 출력된 값들을 astype으로 바꿔주는 구문이 생략된 이유가 궁금했습니다. 출력 값의 type이 np.float이 맞다면 괜찮은 것 같습니다.

File: models/layers.py (hunk 3 above)
Comment (KevinTheRainmaker) [reply models/layers.py:74]: 음 이 부분은 좀더 알아봐야겠네요

File: training.py (hunk 4 above)
Comment (KevinTheRainmaker) [reply training.py:152]: 이 부분도 좀더 알아보겠습니다

File: dataset/dataset.py (hunk 2 above)
Comment (Jeongwook-Lee) [reply dataset/dataset.py:235]: 넵 맞습니다. augmentation output이 np.float여서 생략했습니다.

**Comment by KevinTheRainmaker** [comment]
일단 머지하겠습니다
